Sources_v3/Controllers/MessageControllerGeneric.swift

```diff
@@ -0,0 +1,206 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+import Foundation
+
+public extension Client {
+    func messageController(channelId: ChannelId, messageId: MessageId) -> MessageControllerGeneric<ExtraData> {
+        .init(client: self, channelId: channelId, messageId: messageId)
+    }
+}
+
+public typealias MessageController = MessageControllerGeneric<DefaultDataTypes>
+
+public final class MessageControllerGeneric<ExtraData: ExtraDataTypes>: Controller, DelegateCallable {
+    public let client: Client<ExtraData>
+    public let channelId: ChannelId
+    public let messageId: MessageId
+    
+    public var message: MessageModel<ExtraData>? {
+        state != .inactive ? messageObserver.item : nil
+    }
+    
+    private let environment: Environment
+    
+    private(set) var anyDelegate: AnyMessageControllerDelegate<ExtraData>? {
+        didSet { stateDelegate = anyDelegate }
+    }
+    
+    private lazy var messageObserver = createMessageObserver()
+        .onChange { [unowned self] change in
+            self.delegateCallback {
+                $0?.messageController(self, didChangeMessage: change)
+            }
+        }
+    
+    private lazy var messageUpdater: MessageUpdater<ExtraData> = environment.messageUpdaterBuilder(
+        client.databaseContainer,
+        client.webSocketClient,
+        client.apiClient
+    )
+
+    init(client: Client<ExtraData>, channelId: ChannelId, messageId: MessageId, environment: Environment = .init()) {
+        self.client = client
+        self.channelId = channelId
+        self.messageId = messageId
+        self.environment = environment
+    }
+    
+    public func startUpdating(_ completion: ((Error?) -> Void)? = nil) {
+        do {
+            try messageObserver.startObserving()
+        } catch {
+            callback { completion?(ClientError(with: error)) }
+            return
+        }
+        
+        state = .localDataFetched
+        
+        messageUpdater.getMessage(cid: channelId, messageId: messageId) { [weak self] error in
+            guard let self = self else { return }
+            self.state = error == nil ? .remoteDataFetched : .remoteDataFetchFailed(ClientError(with: error))
+            self.callback { completion?(error) }
+        }
+    }
+}
+
+// MARK: - Actions
+
+public extension MessageControllerGeneric {
+    func editMessage(text: String, completion: ((Error?) -> Void)? = nil) {
```

Comment:
It either should be `edit(` or you should change `delete(` to `deleteMessage(` ... I'll leave that up to you, I don't have preferences.

Reply:
Now it says `deleteMessage`